src/EditorFeatures/Core.Wpf/LineSeparators/LineSeparatorAdornmentManager.cs

```diff
@@ -0,0 +1,117 @@
+            if (removeOldTags)
+            {
+                foreach (var changedSpan in changedSpanCollection)
+                {
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(changedSpan);
+                    }
+                }
+            }
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                // is there any effect on the view?
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
+
+                var tagSpans = TagAggregator.GetTags(changedSpan);
+                foreach (var tagMappingSpan in tagSpans)
+                {
+                    // We don't want to draw line separators if they would intersect a collapsed outlining
+                    // region.  So we test if we can map the start of the line separator up to our visual 
+                    // snapshot. If we can't, then we just skip it.
+                    var point = tagMappingSpan.Span.Start.GetPoint(changedSpan.Snapshot, PositionAffinity.Predecessor);
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    var mappedPoint = TextView.BufferGraph.MapUpToSnapshot(
+                        point.Value, PointTrackingMode.Negative, PositionAffinity.Predecessor, TextView.VisualSnapshot);
+                    if (mappedPoint == null)
+                    {
+                        continue;
+                    }
+
+                    if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, viewSnapshot, out var span))
+                    {
+                        continue;
+                    }
+
+                    if (!viewLines.IntersectsBufferSpan(span))
+                    {
+                        // span is outside of the view so we will not get geometry for it, but may 
+                        // spent a lot of time trying.
+                        continue;
+                    }
+
+                    // add the visual to the adornment layer.
+                    var geometry = viewLines.GetMarkerGeometry(span);
+                    if (geometry != null)
+                    {
+                        var tag = tagMappingSpan.Tag;
+                        var graphicsResult = tag.GetGraphics(TextView, geometry);
+                        AdornmentLayer.AddAdornment(
+                            behavior: AdornmentPositioningBehavior.TextRelative,
+                            visualSpan: span,
+                            tag: tag,
+                            adornment: graphicsResult.VisualElement,
+                            removedCallback: delegate { graphicsResult.Dispose(); });
+                    }
+                }
+            }
```

Comment:
almost all this code felt identical as well.  can we not share almost all (or all) of this in the base class?  what are the ways they need to be different?

Reply:
Some add adornments per line, some add per span. The geometry is also different. Other than they both are adding adornments, I don't see much direct overlap.